### DISCUSSION TITLE
Leader key (a la Vim)

### DISCUSSION THREAD

**Discussion by gerardbosch** [kmonad/kmonad#413] [Ideas]
Hi there! I'd just like to ask if implementing some sort of leader key like VIM could make sense. So actions can be triggered by a sequence of keystrokes in a certain timeout after leader key is tapped.

Some examples:
- `<leader>fnw`  :  Firefow New Window command
- `<leader>jira`  :  Open a new browser window with the Jira URL (so you can quickly create a ticket and close, when the Web browser is not focused, of course).
- `<leader>smi`  :  Emit smiley emoji 🙂 (in case of unicode support)
- `<leader>fp` : 🤦
- `<leader>peace` : ✌(-‿-)✌
- `<leader>shr` :  ¯_(ツ)_/¯
- `<leader>gri` : 😃
- `<leader>foo` : "Any long and repetitive snippet of text you use daily."

This could help to use mnemonics to quickly trigger an "infinite" number of scripts, type stuff,...

I'm not very sure if this feature makes sense for KMonad, or maybe an application launcher would suit this better, but I was wondering about that :)

Cheers!

**Comment by molleweide** [comment]
I myself have been thinking about doing something like this but with a dedicated layer structure so that it becomes like vim leader. If you create lots of layers then you can achieve this and the use shell command aliases for keys.

**Reply by molleweide** [reply]
For example I want to create a layer where I put all my window manager commands.

**Reply by gerardbosch** [reply]
But can you use long pnemonics like `smi` or longer for that? AFAIK, KMonad only lets to launch stuff from a single key.
So, is that possible?

**Reply by molleweide** [reply]
Oh aah good question. I don't know. I was think that you skiss out a smart layer structure so that each key in the sequence switch from one layer to the next. Use one layer as a router for accessing other layers and so on..

**Reply by molleweide** [reply]
It's going to be interesting to see what other people thinks of your qeustion/idea.

**Comment by gerardbosch** [comment]
Hi @david-janssen @slotThe @jokesper, do you think this would fit KMonad? I'm very happy with KMonad and I opened this discussion a few years ago.

I realized that Kanata has support for Vim leader sequences: https://github.com/jtroo/kanata/blob/main/docs/config.adoc#sequences

Do you think leader key fits, or could ever make it into KMonad?

Thank you! 😃

**Reply by jokesper** [reply]
I don't quite remember the exact behavior of `around-next` (or `layer-next`) and I do remember there being an issue about problems with releases, which was fixed by using `around-next-single`. Isn't this pretty much all you need to implement this behavior?
A timeout and an action on key sequence interruption is something which may be missing.
Furthermore modifiers (e.g. when using uppercase keys) would be an interesting experience when using `around-next`

**Reply by gerardbosch** [reply]
Hi @jokesper, to be honest, right now I don't have any idea how this works. The fact that KMonad does not explicitly mention this feature made me think it is not possible. On the other side I recall reading that Kanata added support to this because of missing KMonad (among other reasons).

I'm thinking that would be easy to use it to paste text snippets with simple shell commands like:
`echo 'This is my repetitive text snippet that I directly inject several times a day with a Leader key :)' | clippaste`
(where clippaste is your own shell script or alias depending on your platform).

(I'm tired of typing my email, for example :joy: — you type your email other places than web browser)

In case that was already possible with the current feature set of KMonad, it would be awesome if someone could give a minimal configuration example 😃. And thanks for your awesome work!!

**Reply by jokesper** [reply]
I think something like this:
```clojure
(defcfg
	input (device-file "place-holder")
	output (uinput-sink "E-Mail typer++")

	fallthrough true
)

(defalias
	nl ret ;; newline, you may want to use S-ret instead
)

(defsrc \)
(deflayer default
	(around-next-single (layer-toggle leader))
)

(defsrc :name leader
	f i
	\ ;; To actually press \.
	esc ;; Useful to abort the sequence whithout pressing something
)
(deflayer leader :source leader
	(around-next-single (layer-toggle formal))
	(around-next-single (layer-toggle informal))
	\
	XX
)

(defsrc :name email
	h f
	esc
)
(deflayer formal :source email
	#(H e l l o , @nl @nl )
	#(W i t h
	spc k i n d
	spc r e g a r d s
	@nl @nl
	)
	XX
)

(deflayer informal :source email
	#(S u p , @nl)
	#(B y e @nl)
	XX
)
```
Though keep in mind there is no timeout. And shift+something does not work in sequences (as you would expect it to).
Btw.: to test this I was running kmonad on top of kmonad to keep my keyboard mapping and have the leader key behaviour. Something I would like a better solution for in the future
